src/Api/Orbit.cpp

```diff
@@ -75,4 +76,47 @@ void orbit_api_async_string(const char* str, uint64_t id, orbit_api_color color)
     str += chunk_size;
   }
 }
+
+static void orbit_api_initialize_v0(uint64_t address) {
+  // The api function table is accessed by user code using this pattern:
+  //
+  // if(api->active && api->orbit_api_function_name) api->orbit_api_function_name(...);
+  //
+  // We don't use locking nor memory barriers here for these reasons:
```

Comment:
I have to say, I'm not very convinced, it feels like there should at least be atomics involved. But I've forgotten the little I had learned from this 3-hour talk https://www.youtube.com/watch?v=A8eCGOqgvH4 about the subtleties of this so I don't know.
But in any case: Isn't this code executed while the target is stopped? Or when only one thread is stopped? And does that make a difference?

Reply:
If I'm not mistaken, this is at least undefined behavior and should therefore be avoided. Is there a benefit to avoiding memory barriers here?

Reply:
> But in any case: Isn't this code executed while the target is stopped? Or when only one thread is stopped? 

All threads are stopped. Then one thread gets hijacked for doing the write. Then that thread gets restored. Then all threads are restarted.

>And does that make a difference?

I don't know.

I think the only thing we need to avoid is tearing and I think this does not occur when the data we write is eight byte aligned (no, I don't have a link to the place in the intel manual but I'd love to see one). I'm not so sure if stopping the threads helps with that. I think an unaligned access gets split into two accesses (internal to the processor, not by the compiler, the machine code is still only one instruction) and if the stop happens inbetween (is this possible?) we get tearing. All of this might be wrong!

Reply:
[This looks related](https://docs.microsoft.com/en-us/windows/win32/dxtecharts/lockless-programming?redirectedfrom=MSDN#non-atomic-operations):

_On all modern processors, you can assume that reads and writes of naturally aligned native types are atomic. As long as the memory bus is at least as wide as the type being read or written, the CPU reads and writes these types in a single bus transaction, making it impossible for other threads to see them in a half-completed state. On x86 and x64 there, is no guarantee that reads and writes larger than eight bytes are atomic. This means that 16-byte reads and writes of streaming SIMD extension (SSE) registers, and string operations, might not be atomic._

_Reads and writes of types that are not naturally aligned—for instance, writing DWORDs that cross four-byte boundaries—are not guaranteed to be atomic. The CPU may have to do these reads and writes as multiple bus transactions, which could allow another thread to modify or see the data in the middle of the read or write._

So making sure the pointers are 8-byte aligned in the target should do the trick.

Reply:
Ok so I revisited this to make sure that no one can access the function pointers until the "active" flag has been set.  To achieve this, we use acquire and release semantics in the form of memory fences. It's essentially what is described here (https://preshing.com/20130823/the-synchronizes-with-relation/). Note that since access to naturally aligned uint32_t is atomic, we don't need to make the guard variable a std::atomic<uint32_t> (this was verified with the author of the post).

See also https://preshing.com/20120913/acquire-and-release-semantics/

Reply:
I don't know what the code is as it's not yet updated with fences, but I'm still not sure about not using atomic on the boolean. In C++, a data race, e.g. when one thread writes to the bool and another reads at the same time, is undefined behavior (4.7.1 in the C++17 standard), unless I'm reading something wrong. That is, whether you get correct behavior here is not just a property of the underlying hardware, but also of the compiler (which in the case of undefined behavior may do what it wants). 

What's the benefit of not using an atomic? Performance?

Reply:
std::atomic is equivalent to using load/store with memory_order_seq_cst, which is not necessary in this case, since we only want to make sure that initalized filag is read before function pointer. Note also that atomic_load/store will fail if the value passed is not properly aligned.

Reply:
But isn't there a potential concurrent write and read on the "initialized" or "active" boolean? (Not sure which is the one that is checked by the target process threads.)

Reply:
I am sure they are - but atomic_load/store with acquire/release semantics is not much different from memory_order_seq_cst in that respect, or is it?

Reply:
I'm still not sure what the proposed code will look like. I think to be on the safe side, the code should be something like:

std::atomic<bool> active;

// On thread initializing the function pointers
api->start = &orbit_api_start;
[...]
active.store(true, std::memory_order_release);

// On threads reading the function pointers
if (active.load(std::memory_order_acquire)) {
  // use function pointer
}

Reply:
> What's the benefit of not using an atomic? Performance?
Orbit.h is pure "C" so we don't get the std::atomic, but maybe there is the equivalent in C11? I'll look into it.

> std::atomic is equivalent to using load/store with memory_order_seq_cst
Well in this case we'd use memory_order_relaxed for the guard variable (active)

Reply:
Ah, sorry, I didn't realize that Orbit.h was C only. I'll wait for the code update.

Reply:
I pushed the change. Concerning the fact that Orbit.h is C, there is an additional annoyance that we can't include <stdatomic.h> from C++ (we get compilation errors: https://godbolt.org/z/3nqzM6he7, it seems to be a known issue). 

Since we use this guard variable as a flag, it can be either 0 or 1, even if the write was not atomic, I have a hard time imagining a scenario where that could be an issue. In our case, what is important is that by the time the flag is set to 1, all the function pointers are visible.

Reply:
A non-atomic write would only delay the api enabling. Also, it's worth noting that those function pointers are written to once and never changed again.

Reply:
My (only) point here is that not using atomic on the guard flag is undefined behavior in C++ (and presumably in C11 as well?). But if we can't use https://en.cppreference.com/w/c/language/atomic due to the compilation error, I'm not sure what we can do.

Reply:
Would it be possible and how complicated would it be to switch between the C++ atomics and the C atomics depending on whether it's C++ or C here?
With all the subtleties at compiler and CPU level (more about memory order than tearing) and the undefined behavior I think it would be cleaner and clearer to use atomics.

Reply:
> Would it be possible and how complicated would it be to switch between the C++ atomics and the C atomics depending on whether it's C++ or C here?

Because we access this atomic from C++ in liborbit.so, we'd have to reinterpret the C atomic of Orbit.h as a C++ atomic, which seems risky. The fact that we can't include <stdatomic.h> in C++ is really annoying. I agree that it would be cleaner to simply have an atomic here, but we are dealing with an operation we know to be atomic on our system and pretty much every system:

"For example, it’s common knowledge that on all modern x86, x64, Itanium, SPARC, ARM and PowerPC processors, plain 32-bit integer assignment is atomic as long as the target variable is naturally aligned. You can verify it by consulting your processor manual and/or compiler documentation. In the games industry, I can tell you that a lot of 32-bit integer assignments rely on this particular guarantee." (https://preshing.com/20130618/atomic-vs-non-atomic-operations/#:~:text=For%20example%2C%20it's%20common%20knowledge,target%20variable%20is%20naturally%20aligned.)

Moreover, it seems that in our case, even if the write wasn't atomic, it wouldn't change the fact that whenever the guard variable's value became a potentially torn non-zero value, we'd be sure that the function table was fully visible (this is achieved with the memory barriers).

I'll keep thinking about how we could possibly share an atomic between C and C++, but so far I don't see how to do it cleanly. I'm definitely open to suggestions.

Reply:
https://www.youtube.com/watch?v=A8eCGOqgvH4&t=4779s gives some hint about the fact that the guard variable can be an aligned int. Also, in https://preshing.com/20130922/acquire-and-release-fences/, we see `g_guard.store(1, std::memory_order_relaxed);` . The use of std::memory_order_relaxed here means that we only care about atomicity of the guard variable and not about memory ordering. Memory ordering is taken care of by the acquire and release fences.

I will submit the current implementation.

Reply:
> The use of std::memory_order_relaxed here means that we only care about atomicity of the guard variable and not about memory ordering.

This is fair. As I wrote in my comment, my issue is generally not with possible tearing (thanks to the atomicity that comes from aligned writes), but about memory order, as this can cause incredibly unexpected issues. If we already take care of memory ordering, then I'm happy as long as it's properly documented.

Reply:
> we'd have to reinterpret the C atomic of Orbit.h as a C++ atomic, which seems risky

Yeah yeah, this seems super risky, we should not do that.